src/chainstate/stacks/db/mod.rs

```diff
@@ -517,6 +537,35 @@ pub const MINER_FEE_MINIMUM_BLOCK_USAGE: u64 = 80; // miner must share the first
 
 pub const MINER_FEE_WINDOW: u64 = 24; // number of blocks (B) used to smooth over the fraction of tx fees they share from anchored blocks
 
+pub struct ChainStateBootData {
+    pub first_burnchain_block_hash: BurnchainHeaderHash,
+    pub first_burnchain_block_height: u32,
+    pub first_burnchain_block_timestamp: u32,
+    pub initial_balances: Vec<(PrincipalData, u64)>,
+    pub post_flight_callback: Option<Box<dyn FnOnce(&mut ClarityTx) -> ()>>,
+}
+
+impl ChainStateBootData {
+    pub fn preprend_post_flight_callback(
```

Comment:
Typo: `preprend` -> `prepend`

Reply:
Can you also add a rustdoc here that explains when the callback will be invoked, e.g., after initial balances are inserted, and after boot contracts are initialized, but before any previously inserted post flight callbacks?

Reply:
Honestly, this code scares me.  It makes it difficult to figure out what boot code gets executed, and in what order it gets executed.  This, in turn, makes auditing the boot block more difficult than it needs to be.  I'd much rather keep the code that instantiates the boot block in the same call site as the code that instantiates the chainstate (without any chaining), since it makes it much clearer as to what's supposed to happen when the chainstate is created.

Reply:
I understand the concern, and when I'm also not a fan of this nested callbacks, this construction / pattern is not something that this PR is introducing, it's already present in the `next` branch (https://github.com/blockstack/stacks-blockchain/blob/c0364bcf01a23fb2dfeb31721856e1b23285423e/src/chainstate/coordinator/mod.rs#L256-L288), the only thing that this branch is doing, is encapsulating all the elements in one struct, the ChainStateBootData.
I'm happy to take on this refactoring and see if and how sites can be merged, but I think that it should be addressed in a separate PR, so that we can unblock Xenon deployments now.

Reply:
Comment added in https://github.com/blockstack/stacks-blockchain/pull/1929/commits/6225b4ab7526b9a8454f1efcd759cd345fb6ce22

Reply:
Issue created https://github.com/blockstack/stacks-blockchain/issues/2066

Reply:
Thanks!